### DISCUSSION TITLE
더티 체킹이란 무엇이고, 어떻게 동작하나요?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#157] [Spring Data]
.

**Comment by Kimprodp** [comment]
**[Dirty Checking]**
Dirty Checking은 변경 감지로 트랜잭션이 끝나는 시점에 모든 엔티티 객체의 변경 사항을 감지하고 데이터베이스에 자동으로 반영해주는 기능이다.

**[동작 과정]**
1. flush가 호출되면 영속성 컨텍스트는 Entity의 변경사항을 감지하기 위해 영속성 컨텍스트에 관리되고 있는 모든 Entity를 순회하면서 스냅샷 필드에 잇는 Entity의 상태와 비교한다.
2. 변경 사항이 있는 Entity는 DB에 반영하기 위한 SQL 명령어를 생성하고 쓰기 지연 SQL 저장소에 등록한다.
3. 쓰기 지연 SQL 저장소의 SQL 명령어들이 DB로 전송되어 실행된다. 이때 영속성 컨텍스트와 DB 간의 상태가 동기화 된다. (이 시점에서 트랜잭션이 커밋되지 않았기 때문에 DB에 반영된 것은 아님)
4. 최종적으로 트랜잭션이 커밋되면 flush 과정에서 실행된 SQL 명령어에 의해 변경된 데이터가 실제로 DB에 반영되고 저장된다. 만약 트랜잭션이 롤백되면, flush를 통해 실행된 모든 변경 사항도 롤백된다.

**Comment by seunghye218** [comment]
## Dirty Checking

더티 체킹(Dirty Checking)은 JPA(Java Persistence API)에서 엔티티 객체의 변경 사항을 자동으로 감지하고, 이를 데이터베이스에 반영하는 메커니즘입니다. 코드에서 업데이트를 명시적으로 호출하지 않아도 됩니다.

### 동작 방식

1. 엔티티 관리: JPA 엔티티 매니저(EntityManager)는 영속성 컨텍스트(Persistence Context)에서 엔티티 객체를 관리합니다. 영속성 컨텍스트는 엔티티의 상태를 추적하는 1차 캐시입니다.
2. 스냅샷 생성: 엔티티가 처음 영속성 컨텍스트에 저장될 때, JPA는 해당 엔티티의 현재 상태를 스냅샷으로 저장합니다. 이 스냅샷은 엔티티의 필드 값의 초기 상태를 기록합니다.
3. 변경 감지: 트랜잭션이 진행되는 동안 애플리케이션 코드에서 엔티티 객체의 필드 값을 변경하면, JPA는 트랜잭션 커밋 시점에 엔티티의 현재 상태를 스냅샷과 비교하여 변경 사항을 감지합니다.
4. 변경 사항 반영: 변경 사항이 감지되면, JPA는 자동으로 SQL `UPDATE` 문을 생성하여 변경된 필드 값을 데이터베이스에 반영합니다. 이 과정은 트랜잭션 커밋 시점에 발생합니다.

### 더티 체킹의 이점

1. 개발 생산성 향상: 개발자가 엔티티의 상태 변경에 대한 명시적인 `UPDATE` 쿼리를 작성하지 않아도 됩니다. JPA가 변경 사항을 자동으로 감지하고 데이터베이스에 반영합니다.
2. 유지보수성 향상: 엔티티의 상태 변경 로직이 분리되어 있어 코드가 더 간결하고 이해하기 쉬워집니다. 이는 코드 유지보수성을 높여줍니다.
3. 자동화된 데이터 동기화: 더티 체킹을 통해 데이터베이스와 애플리케이션 상태가 자동으로 동기화되므로, 데이터 일관성을 유지하기가 용이합니다.

### 주의 사항

1. 성능 고려: 더티 체킹은 엔티티의 모든 필드를 스냅샷과 비교하므로, 대량의 엔티티가 변경되는 경우 성능 이슈가 발생할 수 있습니다. 이런 경우 필요한 필드만 업데이트하거나, 변경 감지 전략을 조정하는 등의 최적화가 필요합니다.
2. 명시적 플러시: 변경 사항이 즉시 반영되어야 하는 경우, `EntityManager.flush()` 메서드를 호출하여 영속성 컨텍스트의 변경 사항을 명시적으로 데이터베이스에 반영할 수 있습니다.
3. 트랜잭션 관리: 더티 체킹은 트랜잭션 커밋 시점에 발생하므로, 트랜잭션 관리가 중요합니다. 트랜잭션 범위를 명확히 설정하여 데이터 일관성을 유지해야 합니다.

**Comment by le2sky** [comment]
엔티티의 변경사항을 데이터베이스에 자동으로 반영하는 기능을 변경 감지라 한다.
- 트랜잭션 커밋 시, em 내부에서 플러시가 호출된다. 
- 엔티티와 스냅샷(영컨 보관 최초 상태)을 비교해서 변경된 엔티티를 찾는다.
- 변경된 엔티티가 있다면, 수정 쿼리를 생성하여 쓰기 지연 저장소에 보낸다.
- 쓰기 지연 저장소의 SQL을 데이터베이스에 보낸다.
- DB 트랜잭션을 커밋한다.
